server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java

```diff
@@ -877,22 +877,21 @@ public void commit(CommitSession commitSession, List<Mutation> mutations) {
       totalBytes += mutation.numBytes();
     }
 
-    getTabletMemory().mutate(commitSession, mutations, totalCount);
-
-    synchronized (this) {
-      if (isCloseComplete()) {
-        throw new IllegalStateException(
-            "Tablet " + extent + " closed with outstanding messages to the logger");
-      }
-      // decrement here in case an exception is thrown below
-      decrementWritesInProgress(commitSession);
-
+    try {
+      getTabletMemory().mutate(commitSession, mutations, totalCount);
       getTabletMemory().updateMemoryUsageStats();
-
       numEntries += totalCount;
       numEntriesInMemory += totalCount;
       ingestCount += totalCount;
       ingestBytes += totalBytes;
```

Comment:
The increments needs to be done in in sync block and maybe update of stats.  So would do something like the following.  The exception check would be best done outside the finally block, in the case where an exception is thrown and the finally block is running we do not want throw a new exception as it causes the old one to be lost.  Also validation is only needed when mutate was successful, if the mutate failed for some reason we do not need to do the validaiton.

```suggestion
synchronized (this) {
       getTabletMemory().updateMemoryUsageStats();
        if (isCloseComplete()) {
          throw new IllegalStateException(
              "Tablet " + extent + " closed with outstanding messages to the logger");
        }
      numEntries += totalCount;
      numEntriesInMemory += totalCount;
      ingestCount += totalCount;
      ingestBytes += totalBytes;
      }
```

Reply:
ok, for some reason I misunderstood your previous suggestion.